iree/testing/vulkan/iree-run-module-vulkan-gui-main.cc

```diff
@@ -0,0 +1,452 @@
+static VkDevice g_Device = VK_NULL_HANDLE;
+static uint32_t g_QueueFamily = (uint32_t)-1;
+static VkQueue g_Queue = VK_NULL_HANDLE;
+static VkPipelineCache g_PipelineCache = VK_NULL_HANDLE;
+static VkDescriptorPool g_DescriptorPool = VK_NULL_HANDLE;
+
+static ImGui_ImplVulkanH_Window g_MainWindowData;
+static uint32_t g_MinImageCount = 2;
+static bool g_SwapChainRebuild = false;
+static int g_SwapChainResizeWidth = 0;
+static int g_SwapChainResizeHeight = 0;
+
+namespace iree {
+namespace {
+
+void check_vk_result(VkResult err) {
+  if (err == 0) return;
+  IREE_LOG(FATAL) << "VkResult: " << err;
+}
+
+void CleanupVulkan() {
+  vkDestroyDescriptorPool(g_Device, g_DescriptorPool, g_Allocator);
+
+  vkDestroyDevice(g_Device, g_Allocator);
+  vkDestroyInstance(g_Instance, g_Allocator);
+}
+
+void CleanupVulkanWindow() {
+  ImGui_ImplVulkanH_DestroyWindow(g_Instance, g_Device, &g_MainWindowData,
+                                  g_Allocator);
+}
+
+StatusOr<std::string> GetModuleContentsFromFlags() {
+  auto module_file = absl::GetFlag(FLAGS_module_file);
+  std::string contents;
+  if (module_file == "-") {
+    contents = std::string{std::istreambuf_iterator<char>(std::cin),
+                           std::istreambuf_iterator<char>()};
+  } else {
+    IREE_ASSIGN_OR_RETURN(contents, file_io::GetFileContents(module_file));
+  }
+  return contents;
+}
+
+// Runs the current IREE bytecode module and renders its result to a window
+// using ImGui.
+Status ImGuiRender(iree_hal_device_t* device, iree_vm_context_t* context,
+                   iree_vm_function_t function) {
+  ImGui::Begin(absl::GetFlag(FLAGS_module_file).c_str(), /*p_open=*/nullptr,
+               ImGuiWindowFlags_AlwaysAutoResize);
+
+  IREE_RETURN_IF_ERROR(ValidateFunctionAbi(function));
+
+  IREE_ASSIGN_OR_RETURN(auto input_descs, ParseInputSignature(function));
+  vm::ref<iree_vm_list_t> inputs;
+  if (!absl::GetFlag(FLAGS_inputs_file).empty()) {
+    if (!absl::GetFlag(FLAGS_inputs).empty()) {
+      return InvalidArgumentErrorBuilder(IREE_LOC)
+             << "Expected only one of inputs and inputs_file to be set";
+    }
+    IREE_ASSIGN_OR_RETURN(
+        inputs, ParseToVariantListFromFile(input_descs,
+                                           iree_hal_device_allocator(device),
+                                           absl::GetFlag(FLAGS_inputs_file)));
+  } else {
+    IREE_ASSIGN_OR_RETURN(
+        inputs,
+        ParseToVariantList(input_descs, iree_hal_device_allocator(device),
+                           absl::GetFlag(FLAGS_inputs)));
+  }
+
+  IREE_ASSIGN_OR_RETURN(auto output_descs, ParseOutputSignature(function));
+  vm::ref<iree_vm_list_t> outputs;
+  IREE_RETURN_IF_ERROR(iree_vm_list_create(/*element_type=*/nullptr,
+                                           output_descs.size(),
+                                           iree_allocator_system(), &outputs));
+
+  std::string function_name = absl::GetFlag(FLAGS_entry_function);
```

Comment:
File IO, flag queries, and other one-time initialization code should be moved out of the update loop.

Reply:
Yeah good point. Moved out.